F5 gtm support

dnapper opened the pull request:
DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

Adds discovery and polling of F5 GTM module Wide IP and Pools discovery.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/42e9b0ae-4668-4025-906b-cd84468de24e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/42e9b0ae-4668-4025-906b-cd84468de24e) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/7e48572f-b52d-4931-a5c1-d8d9f23870dd) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/7e48572f-b52d-4931-a5c1-d8d9f23870dd) completed: **No new issues**

laf commented:
Thanks for this. Before we review the code can you share some screenshots of what the various new pages look like please.

Could you also show the impact on poll time for your devices and indicate roughly how big they are? The f5 module already can have a big impact on users so we need to understand the risk to existing installs here.

dnapper commented:
Sure. 

In our environment we don't run GTM services on the same devices as we run LTM so for me it will have no impact on poll times on existing devices. 

The code checks to see if the GTM module is running prior to discovering any new entities so I don't expect this will impact any devices that have the module disabled. 

This adds two new tabs to the load balancer page if GTM objects are discovered. 
![gtm-newtabs](https://user-images.githubusercontent.com/29460571/35537753-9c9bab84-0508-11e8-89c0-7820577d1f1e.PNG)

The pages look virtually the same as the current VS and Pool pages but the graphs are based on DNS requests handled not data throughput. 
![gtm-wideipdetail](https://user-images.githubusercontent.com/29460571/35537747-94828f1c-0508-11e8-883b-559ca7dbb853.PNG)
![gtm-wideipdetail](https://user-images.githubusercontent.com/29460571/35537782-b9757424-0508-11e8-8aeb-89235f3ac8b6.PNG)

I'll leave these pre devices with this running for a few days so you can see what the data looks like (and that it's actually polling :) )

One thing i've been struggling with is how to handle representing a service that has multiple pools associated.  In the end I decided that if I can sort out a way I can always add that later. 

This is with a pre-prod devices with ~30 objects on it. 

While on the main branch the discovery time and polling time is ~7s. 
Discovered in 7.116 seconds
SNMP [45/5.64s]: Get[10/0.77s] Getnext[1/0.07s] Walk[34/4.80s]
MySQL [158/0.06s]: Cell[30/0.00s] Row[25/0.00s] Rows[70/0.01s] Column[1/0.00s] Update[26/0.03s] Insert[5/0.01s] Delete[1/0.00s]
RRD [0/0.00s]: Update[0/0.00s] Create [0/0.00s] Other[0/0.00s]


Polled in 7.028 seconds
UPDATED!
#### Start Alerts ####
#### End Alerts ####
SNMP [57/5.61s]: Get[19/1.41s] Getnext[0/0.00s] Walk[38/4.20s]
MySQL [63/0.07s]: Cell[3/0.00s] Row[2/0.00s] Rows[20/0.00s] Column[3/0.00s] Update[33/0.06s] Insert[2/0.00s] Delete[0/0.00s]
RRD [75/0.00s]: Update[75/0.00s] Create [0/0.00s] Other[0/0.00s]

With the new additions discovery takes ~10s
Discovered in 10.139 seconds

SNMP [49/8.12s]: Get[11/0.87s] Getnext[1/0.07s] Walk[37/7.17s]
MySQL [704/0.60s]: Cell[30/0.01s] Row[26/0.00s] Rows[73/0.01s] Column[1/0.00s] Update[86/0.09s] Insert[487/0.48s] Delete[1/0.00s]
RRD [0/0.00s]: Update[0/0.00s] Create [0/0.00s] Other[0/0.00s]

Polling takes ~8s

Polled in 8.377 seconds
UPDATED!
#### Start Alerts ####
#### End Alerts ####
SNMP [64/6.76s]: Get[19/1.40s] Getnext[0/0.00s] Walk[45/5.36s]
MySQL [65/0.07s]: Cell[3/0.00s] Row[2/0.00s] Rows[22/0.01s] Column[3/0.00s] Update[33/0.06s] Insert[2/0.00s] Delete[0/0.00s]
RRD [195/0.01s]: Update[135/0.00s] Create [60/0.00s] Other[0/0.00s]

dnapper commented:
Update:
I found an issue where once I added an LTM to my test server the LTM discovery process was deleting the GTM objects. In order to fix it I split the GTM polling into it's own load balancer module and it appears to be working.

I'll leave it for a while to see if the problem comes back.

As a side note here is a comparison of LTM poling before and after the patch.
The polling time is quite a bit faster on my dev server but that's probably due to load.

DEV
![polling-dev](https://user-images.githubusercontent.com/29460571/35571135-7e30f754-058e-11e8-9e7e-1a13dd95c298.PNG)

PRE
![polling-pre](https://user-images.githubusercontent.com/29460571/35571270-d4e61fac-058e-11e8-8b24-38e1aaee7dcb.PNG)

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/06f7ea99-9b65-4f34-97de-7ee0a00a4c8a) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/9d3f33ba-5f9c-4c00-b567-14061d8eb5e0) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/06f7ea99-9b65-4f34-97de-7ee0a00a4c8a) completed: **No new issues**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/9d3f33ba-5f9c-4c00-b567-14061d8eb5e0) completed: **No new issues**

laf commented on html/pages/device/loadbalancer/gtm_pool.inc.php:
```diff
@@ -0,0 +1,39 @@
+}
+$components = $keep;
+
+global $config;
```
You shouldn't need this.

laf commented on html/pages/device/loadbalancer/gtm_pool.inc.php:
```diff
@@ -0,0 +1,39 @@
+
+global $config;
+
+if (is_file('pages/device/loadbalancer/'.mres($vars['subtype']).'.inc.php')) {
```
You can drop the mres from these two lines, `$vars` has already had that applied.

dnapper replied:
Do you want me to remove it from the ltm module pages as well?

laf replied:
No, leave that. You can remove them in another PR but let's not mix things up here.

laf commented on html/pages/device/loadbalancer/gtm_wide.inc.php:
```diff
@@ -0,0 +1,39 @@
+
+global $config;
+
+if (is_file('pages/device/loadbalancer/'.mres($vars['subtype']).'.inc.php')) {
```
You don't need to use mres on these two lines.

laf commented on includes/discovery/loadbalancers/f5-gtm.inc.php:
```diff
@@ -0,0 +1,187 @@
+ * null  == timeout or something else that caused an error, OID may exist but we couldn't get it.
+ */
+
+if (!is_null($gtmWideIPEntry) || !is_null($gtmWideStatusEntry) || !is_null($gtmPoolEntry)) {
```
Don't these need to be && if you want them all to match?

dnapper replied:
You don't need to have a pool attached to a wide IP for it to function so there could be situations where people are using the device to service static DNS records as well as dynamic ones based on pool conditions.

laf commented on includes/discovery/loadbalancers/f5-gtm.inc.php:
```diff
@@ -0,0 +1,187 @@
+
+//Check if GTM modeul is running and run object discovery.
+if ((snmp_get($device, 'sysModuleAllocationProvisionLevel.3.103.116.109', '-Ovqs', 'F5-BIGIP-SYSTEM-MIB')) !=none) {
+    $gtmWideIPEntry = snmpwalk_array_num($device, '1.3.6.1.4.1.3375.2.3.12.1.2.1', 0);
```
If we need all of these walks then it would be worth checking if each contains the data before trying the next and only carry on if we have them all (moving the check from below to here).

dnapper replied:
Sure, there is no point moving onto look for status and pools if there are no wide IPs. 
Do you want me to do the same in the LTM discovery? We could skip trying to get pools and status codes if there are no virtual servers objects.

laf replied:
Any saving is good, but again I'd keep it to another PR for LTM.

dnapper replied:
Can do.

laf commented on includes/discovery/loadbalancers/f5-ltm.inc.php:
```diff
@@ -51,6 +51,13 @@
 $ltmPoolMemberEntry = snmpwalk_array_num($device, '1.3.6.1.4.1.3375.2.2.5.3.2.1', 0);
 $ltmPoolMbrStatusEntry = snmpwalk_array_num($device, '1.3.6.1.4.1.3375.2.2.5.6.2.1', 0);
 
+//Check if GTM modeul is running and run object discovery.
```
Is this supposed to be here?

dnapper replied:
no... whoops.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/06a5a4c1-7120-429c-aad9-f071e8c04c79) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/librenms/librenms/inspections/06a5a4c1-7120-429c-aad9-f071e8c04c79) completed: **No new issues**

murrant commented:
Looks good, thanks @dnapper :)
